readme.md
# 🖥️ HTML2PDF
An easy-to-use but reliable asynchronous library to create highly customizable PDFs from HTML or URL as ``buffer``, ``base64 string`` and ``.pdf file``. Other than most npm-solutions for html-to-pdf conversion, this one leverages ``puppeteer`` as opposed to the deprecated and unmaintained ``phantomJS``. It also provides far more options to configure to your liking than most similar solutions. This *might* be the best html-to-pdf solution on npm at this point.

## Installation
```
npm i [-g] better-html-pdf
```

## Available functions
Other than the main converter function, HTML2PDF exposes two more functions that allow writing pdf file content as both ``base64`` or ``buffer``. Those functions are also used internally to replace the situationally unreliable ``puppeteer``-native file save function that is controlled through the ``path`` option.

### Function signatures
Convert html to pdf
```typescript
html2pdf(html: string, options: Object) : Promise<string | Buffer>
```
Convert base64 file content to PDF file
```typescript
base64ToPdf(base64: string, file: string) : void
```
Convert file content buffer to PDF file
```typescript
bufferToPdf(buffer: Buffer, file: string) : void
```

### Usage
Using HTML2PDF is quick and simple: Import the ``html2pdf`` function and pass it the html and options.
```javascript
const { html2pdf, base64ToPdf, bufferToPdf } = require('better-html-pdf');

//[...]
// convert html to PDF
const fileContentB64 = await html2pdf('<h1>Test</h1>', { avoidTableRowBreak: true, marginTop: 10, repeatTableHeader: false });
const fileContentBuffer = await html2pdf('<h1>Test</h1>', { fileType: 'buffer', url: 'https://google.com/', viewPort: '1000x700' });

//convert output to pdf file
base64ToPdf(fileContentB64, './test1.pdf');
bufferToPdf(fileContentBuffer, './test2.pdf');
```

Full type declarations are included so ES6 imports are available, too.
```javascript
import { html2pdf } from 'better-html-pdf';

//[...]
// convert html to PDF
const fileContentB64 = await html2pdf('<h1>Test</h1>', { avoidTableRowBreak: true, marginTop: 10, repeatTableHeader: false });
```
### Options
This solution provides a great number of options to configure for your conversion, passed as a javascript object.
**Find a detailed doc of all options and defaults here**: [Options Documentation](https://github.com/0tii/html-to-pdf/blob/master/options.md).

```javascript
{
    fileType: 'base64', // 'base64' | 'buffer'
    url: '', //if specified ignores html
    viewPort: '1920x1080', //string (width)x(height)
    timeout: 5000, //timeout for page loading in ms
    landscape: false,
    format: '', //letter | legal | tabloid | ledger | a0 | a1 | a2 | a3 | a4 | a5 | a6
    repeatTableHeader: true, //repeat html table headers on each page - note: headers only repeat when in <thead>
    repeatTableFooter: true, //repeat html table footers on each page - note: footers only repeat when in <tfoot>
    displayHeaderFooter: true,
    headerTemplate: '',
    footerTemplate: '',
    width: '1920', //document size in pixels or with units (in, mm, cm)
    height: '1080', //document size in pixels or with units (in, mm, cm)
    marginTop: 0, //num in pixels or with units (in, mm, cm)
    marginBottom: 0, //num in pixels or with units (in, mm, cm)
    marginLeft: 0, //num in pixels or with units (in, mm, cm)
    marginRight: 0, //num in pixels or with units (in, mm, cm)
    breakImages: false, //break images between pages
    avoidTableRowBreak: true, //tries avoiding breaking table rows between pages
    avoidDivBreak: false, //tries to avoid breaking divs between pages - can cause unwanted behavior
    omitBackground: false, //hide html background, allows for transparency
    pageRanges: '', //'1-12' | '3-5'
    path: '', //file save path, if empty no file is created
    disableJavascript: false, //disable javascript on the target site/html
    preferCSSPageSize: false, //css-declared page size takes precedent over format, width and height
    printBackground: true, //apply background styling
    trueColors: true, //use unaltered colors
    scale: 1, //render scale, must be between 0.1 and 2
    screenMedia: false //use 'screen' instead of 'print' CSS media
}
```

## Dependencies
HTML2PDF depends on `puppeteer`.

CHANGELOG
# HTML2PDF © 2022 Daniel Rauhut

[0.0.1] 
- Initial release

[0.0.2]
- Added generic ts declaration in order to allow module import

[0.0.3]
- Added proper typings

[0.0.4]
- Readme corrected to include valid npm install command

[0.1.0]
- Typed parameters homogenity
- Fix html2pdf function signature to return Promise instead of the type it resolves to

[0.1.1]
- Detailed documentation for all available options

[0.1.2]
- Improved documentation

[0.1.3]
- Export type declaration for FileType

[0.1.4]
- Republish

test.js
const { html2pdf, base64ToPdf } = require('./index.js');
const { html } = require('./html.js');

async function test() {
    const b64 = await html2pdf(html, { avoidTableRowBreak: true, marginTop: 10, repeatTableHeader: false, landscape: true }); //path must exist
    base64ToPdf(b64, './test.pdf');
}

test();

index.d.ts
declare module 'better-html-pdf';

/**
 * Convert a html string or website to a pdf file
 * @param {*} html html string
 * @param {*} options options object, defaults are respected
 * @returns PDF file content as ``[base64 | buffer]``
 */
export declare function html2pdf(html: string, options: PdfOptions): Promise<string> | Promise<Buffer>;

/**
 * Write a base64 encoded pdf to a pdf file. Path must exist.
 * @param {*} fileContent pdf file content as b64 string
 * @param {*} file path to write the file to, absolute or relative
 */
export declare function base64ToPdf(fileContent: string, file: string): void;

/**
 * Write a buffer to a pdf file. Path must exist.
 * @param {*} fileContent pdf file content as buffer
 * @param {*} file path to write the file to, absolute or relative
 */
export declare function bufferToPdf(fileContent: Buffer, file: string): void;

/**
 * All options that can be passed to the converter
 */
export declare interface PdfOptions {
    fileType?: FileType,
    url?: string,
    viewPort?: string,
    timeout?: number,
    landscape?: boolean,
    format?: PageFormat,
    repeatTableHeader?: boolean,
    repeatTableFooter?: boolean,
    displayHeaderFooter?: boolean,
    headerTemplate?: string,
    footerTemplate?: string,
    width?: number | string,
    height?: number | string,
    marginTop?: number | string,
    marginBottom?: number | string,
    marginLeft?: number | string,
    marginRight?: number | string,
    breakImages?: boolean,
    avoidTableRowBreak?: boolean,
    avoidDivbreak?: boolean,
    omitBackground?: boolean,
    pageRanges?: string,
    path?: string,
    disableJavascript?: boolean,
    preferCSSPageSize?: boolean,
    printBackground?: boolean,
    trueColors?: boolean,
    scale?: number,
    screenMedia?: boolean
}

export declare enum PageFormat{
    letter = 'letter',
    legal = 'legal',
    tabloid = 'tabloid',
    a0 = 'a0',
    a1 = 'a1',
    a2 = 'a2',
    a3= 'a3',
    a4 = 'a4'
}

export type FileType = 'buffer' | 'base64';

index.js
/*
Simple asynchronous library to convert html or URL to PDF stream by leveraging puppeteer.0

© 2022 Daniel Rauhut
*/
const puppeteer = require('puppeteer');
const fs = require('fs');

/**
 * Convert a html string or website to a pdf file
 * @param {*} html html string
 * @param {*} options options object, refer to the doc for specific options or 
 * @see convertWithDefaults
 * @returns HTML converted to PDF as either base64 string or buffer
 */
async function convert(html, options) {
    //add margins in puppeteer format
    options['margin'] = {
        bottom: options.marginBottom,
        left: options.marginLeft,
        right: options.marginRight,
        top: options.marginTop
    }

    return new Promise(async (resolve, reject) => {
        const browser = await puppeteer.launch({ headless: true });

        const page = await browser.newPage();

        try {
            await page.setViewport({
                width: parseInt(options.viewPort.split('x')[0]),
                height: parseInt(options.viewPort.split('x')[1]),
            });
        } catch {
            reject('Bad Viewport option.');
        }

        if (options.disableJavascript)
            await page.setJavaScriptEnabled(false);

        try {
            //navigate to or create desired content
            if (options.url) {
                await page.goto(options.url, {
                    WaitUntil: 'domcontentloaded',
                    timeout: options.timeout
                });
            }
            else {
                await page.setContent(html, {
                    waitUntil: 'domcontentloaded',
                    timeout: options.timeout
                });
            }

            await addCSSOptions(page, options);

            if (options.screenMedia)
                await page.emulateMediaType('screen');

            //check for path option and save manually to avoid unreliable puppeteer file save
            if (options.path) {
                var path = options.path;
                options.path = '';
            }

            const pdfStream = await page.createPDFStream(options);

            const pdfBuffer = await streamToBuffer(pdfStream);

            if (path)
                bufferToPdf(pdfBuffer, path);

            await browser.close();

            if (options.fileType === 'buffer')
                resolve(pdfBuffer);
            else
                resolve(pdfBuffer.toString('base64'));

        } catch (err) {
            reject(err);
        }
    });
}

/**
 * Processes options that are based on injecting CSS into the target page
 * @param {*} page 
 * @param {*} options 
 */
async function addCSSOptions(page, options) {
    //inject css to repeat table header/footer on each page
    if (options.repeatTableHeader === false)
        await page.addStyleTag({ content: `thead {display: table-row-group; }` });

    if (options.repeatTableFooter === false)
        await page.addStyleTag({ content: `tfoot {display: table-row-group; }` });

    //inject css for accurate colors
    if (options.trueColors)
        await page.addStyleTag({ content: `html, body { print-color-adjust: exact }` });

    //inject css to avoid rows breaking between pages
    if (options.avoidTableRowBreak) {
        await page.addStyleTag({ content: `table { break-inside:auto }` });
        await page.addStyleTag({ content: `tr { break-inside:avoid; break-after:auto }` });
    }

    //inject css to avoid breaking images
    if (!options.breakImages) 
        await page.addStyleTag({ content: `img { break-inside: avoid }` });

    //inject css to avoid divs breaking between pages
    if (options.avoidDivBreak) 
        await page.addStyleTag({ content: `div { break-inside:avoid }` });
}

/**
 * Create a buffer from a stream
 * @param {*} stream the file stream
 * @returns buffer of stream content
 */
async function streamToBuffer(stream) {
    const chunks = [];
    return new Promise((resolve, reject) => {
        stream.on('data', (chunk) => chunks.push(Buffer.from(chunk)));
        stream.on('error', (err) => reject(err));
        stream.on('end', () => resolve(Buffer.concat(chunks)));
    });
}

/**
 * write a base64 encoded pdf to a pdf file. Path must exist.
 * @param {*} base64 pdf file content as b64 string
 * @param {*} file path to write the file to, absolute or relative
 */
function base64ToPdf(fileContent, file) {
    const buffer = Buffer.from(fileContent, 'base64');
    bufferToPdf(buffer, file);
}

/**
 * write a buffer to a pdf file. Path must exist.
 * @param {*} fileContent pdf file content as buffer
 * @param {*} file path to write the file to, absolute or relative
 */
async function bufferToPdf(fileContent, file) {
    fs.writeFileSync(file, fileContent);
}

//? vvv this is a wrapper to keep the convert function clean while keeping option defaults in one place. vvv

/**
 * Convert a html string or website to a pdf file
 * @param {*} html html string
 * @param {*} options options object, defaults are respected
 * @returns HTML converted to PDF as either base64 string or buffer
 */
exports.html2pdf = async function html2pdf(
    html,
    {
        fileType = 'base64',
        url = '',
        viewPort = '1920x1080',
        timeout = 5000,
        landscape = false,
        format = '',
        repeatTableHeader = false,
        repeatTableFooter = false,
        displayHeaderFooter = true,
        headerTemplate = '',
        footerTemplate = '',
        width = '1920',
        height = '1080',
        marginTop = 0,
        marginBottom = 0,
        marginLeft = 0,
        marginRight = 0,
        breakImages = false,
        avoidTableRowBreak = true,
        avoidDivBreak = false,
        omitBackground = false,
        pageRanges = '',
        path = '',
        disableJavascript = false,
        preferCSSPageSize = false,
        printBackground = true,
        trueColors = true,
        scale = 1,
        screenMedia = false
    } = {}
) {
    return await convert(
        html,
        {
            fileType: fileType,
            url: url,
            viewPort: viewPort,
            timeout: timeout,
            landscape: landscape,
            format: format,
            repeatTableHeader: repeatTableHeader,
            repeatTableFooter: repeatTableFooter,
            displayHeaderFooter: displayHeaderFooter,
            headerTemplate: headerTemplate,
            footerTemplate: footerTemplate,
            width: width,
            height: height,
            marginTop: marginTop,
            marginBottom: marginBottom,
            marginLeft: marginLeft,
            marginRight: marginRight,
            breakImages: breakImages,
            avoidTableRowBreak: avoidTableRowBreak,
            avoidDivbreak: avoidDivBreak,
            omitBackground: omitBackground,
            pageRanges: pageRanges,
            path: path,
            disableJavascript: disableJavascript,
            preferCSSPageSize: preferCSSPageSize,
            printBackground: printBackground,
            trueColors: trueColors,
            scale: scale,
            screenMedia: screenMedia
        }
    );
}
exports.base64ToPdf = base64ToPdf;
exports.bufferToPdf = bufferToPdf;